di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/utils/HandlerHelper.java

```diff
@@ -453,26 +449,100 @@ private Map<String, Object> convertToMap(Object input) {
+            JWEHeader.Builder headerBuilder =
+                    new JWEHeader.Builder(JWEAlgorithm.RSA_OAEP_256, EncryptionMethod.A256GCM)
+                            .contentType("JWT");
+
+            List<WellKnownJwk> jwkList;
+            String hashedKeyID = "";
+            RSAKey rsaKey;
+            boolean useKeyRotation = credentialIssuer.useKeyRotation();
+            LOGGER.info("useKeyRotation is set to {}", useKeyRotation);
+            if (useKeyRotation) {
+                // automated key rotation path
+
+                jwkList = getWellKnownJwks(credentialIssuer);
+
+                int jwkListLast = jwkList.size() - 1;
+                // Gets current public encryption key
+                WellKnownJwk wellKnownJwk = jwkList.get(jwkListLast);
+                rsaKey = new RSAKey.Builder(wellKnownJwk.getN(), wellKnownJwk.getE()).build();
+                String encryptionKeyId = wellKnownJwk.getKid();
+
+                hashedKeyID = getHashedKeyId(encryptionKeyId);
+                // passes hashed current encryption key ID in JWE header
+                headerBuilder.keyID(hashedKeyID);
+
+            } else {
+
+                String keyValueFromFile = credentialIssuer.publicEncryptionJwkBase64();
+                rsaKey = getEncryptionPublicKey(keyValueFromFile);
+            }
+
+            JWEObject jweObject = new JWEObject(headerBuilder.build(), new Payload(signedJWT));
+
+            encryptJWEObject(jweObject, rsaKey);
 
             return EncryptedJWT.parse(jweObject.serialize());
-        } catch (JOSEException | java.text.ParseException e) {
-            throw new RuntimeException("JWT encryption failed", e);
+        } catch (java.text.ParseException e) {
+            LOGGER.error("Error parsing JWT");
+            throw new RuntimeException(e);
         }
     }
 
-    private RSAKey getEncryptionPublicKey(CredentialIssuer credentialIssuer)
-            throws java.text.ParseException {
-        return RSAKey.parse(
-                new String(
-                        Base64.getDecoder().decode(credentialIssuer.publicEncryptionJwkBase64())));
+    private String getHashedKeyId(String encryptionKeyId) {
+        String hashedKeyID;
+        try {
+            hashedKeyID = createHashedKeyId(encryptionKeyId);
+            LOGGER.info("Key ID successfully hashed");
+        } catch (NoSuchAlgorithmException e) {
+            LOGGER.error("Error: Required key hashing algorithm not found");
+            throw new RuntimeException(e);
+        }
+        return hashedKeyID;
+    }
+
+    // creates hashed key ID from current encryption key ID obtained from JWKS Endpoint
+    private String createHashedKeyId(String keyId) throws NoSuchAlgorithmException {
+        MessageDigest digest = MessageDigest.getInstance("SHA-256");
+        byte[] hash = digest.digest(keyId.getBytes(StandardCharsets.UTF_8));
+        return Hex.encodeHexString(hash);
```

Comment:
I did'nt relise the RFC example needed a new dep todo this.

Can you log out the value and test that this would do the sample.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/HexFormat.html

If so can you then remove the commons codec dep.